File: problems/07_trees/0199_binary_tree_right_side_view/solution.py
"""
Problem: 0199. Binary Tree Right Side View
Difficulty: Medium
URL: https://leetcode.com/problems/binary-tree-right-side-view/

解法：BFS 取每層最後節點值，O(n)/O(w)。
"""
from __future__ import annotations
from typing import Optional, List
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right


def from_level(arr: List[Optional[int]]) -> Optional[TreeNode]:
    if not arr: return None
    root = TreeNode(arr[0]); q = deque([root]); i = 1
    while q and i < len(arr):
        n = q.popleft()
        if i < len(arr) and arr[i] is not None: n.left = TreeNode(arr[i]); q.append(n.left)
        i += 1
        if i < len(arr) and arr[i] is not None: n.right = TreeNode(arr[i]); q.append(n.right)
        i += 1
    return root


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []
        out = []
        q = deque([root])
        while q:
            sz = len(q)
            for i in range(sz):
                n = q.popleft()
                if i == sz - 1:
                    out.append(n.val)
                if n.left: q.append(n.left)
                if n.right: q.append(n.right)
        return out


if __name__ == "__main__":
    sol = Solution()
    assert sol.rightSideView(from_level([1, 2, 3, None, 5, None, 4])) == [1, 3, 4]
    assert sol.rightSideView(from_level([1, None, 3])) == [1, 3]
    assert sol.rightSideView(from_level([])) == []
    assert sol.rightSideView(from_level([1, 2])) == [1, 2]
    assert sol.rightSideView(from_level([1, 2, 3, 4])) == [1, 3, 4]
    print("All tests passed!")

File: problems/07_trees/0199_binary_tree_right_side_view/solution.dart
// Problem: 0199. Binary Tree Right Side View
// Difficulty: Medium

class TreeNode {
  int val; TreeNode? left, right;
  TreeNode([this.val = 0, this.left, this.right]);
}

TreeNode? fromLevel(List<int?> a) {
  if (a.isEmpty) return null;
  final root = TreeNode(a[0]!);
  final q = <TreeNode>[root]; var qi = 0, i = 1;
  while (qi < q.length && i < a.length) {
    final n = q[qi++];
    if (i < a.length && a[i] != null) { n.left = TreeNode(a[i]!); q.add(n.left!); }
    i++;
    if (i < a.length && a[i] != null) { n.right = TreeNode(a[i]!); q.add(n.right!); }
    i++;
  }
  return root;
}

List<int> rightSideView(TreeNode? root) {
  if (root == null) return [];
  final out = <int>[];
  var q = <TreeNode>[root];
  while (q.isNotEmpty) {
    final nx = <TreeNode>[];
    for (final n in q) {
      if (n.left != null) nx.add(n.left!);
      if (n.right != null) nx.add(n.right!);
    }
    out.add(q.last.val);
    q = nx;
  }
  return out;
}

void _expect(List<int> a, List<int> e, String msg) {
  if (a.length != e.length || !List.generate(a.length, (i) => a[i] == e[i]).every((x) => x)) {
    throw StateError('FAIL $msg: $a vs $e');
  }
}

void main() {
  _expect(rightSideView(fromLevel([1, 2, 3, null, 5, null, 4])), [1, 3, 4], 'ex1');
  _expect(rightSideView(fromLevel([1, null, 3])), [1, 3], 'ex2');
  _expect(rightSideView(fromLevel([])), [], 'empty');
  _expect(rightSideView(fromLevel([1, 2])), [1, 2], 'left only');
  _expect(rightSideView(fromLevel([1, 2, 3, 4])), [1, 3, 4], 'mixed');
  print('All tests passed!');
}
